### PR TITLE
nixos/nextcloud: Add option for using object storage as primary storage

### PR DESCRIPTION
**Pull request by mitchmindtree**
This allows to declaratively configure an S3 class object storage as the
primary storage for the nextcloud service. Previously, this could only
be achieved by manually editing the `config.php`.

I've started testing this today with my own digitalocean nextcloud
instance, which now points to my digitalocean S3-compatible "Space" and
all appears to be working smoothly.

My motivation for this change is my recent discovery of how much cheaper
some S3-compatible object storage options are compared to digitalocean's
"Volume" options.

Implementation follows the "Simple Storage Service" instructions here:

https://docs.nextcloud.com/server/latest/admin_manual/configuration_files/primary_storage.html

I have neglected to implement a submodule for the OpenStack Swift
object storage as I don't personally have a use case for it or a method
to test it, however the new `nextcloud.objectstore.s3` submodule should
act as a useful guide for anyone who does wish to implement it.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (aanderse) [inline nixos/modules/services/web-apps/nextcloud.nix:330]: Please remove this option.

**Review by aanderse — CHANGES_REQUESTED** [review]
Will defer to @Ma27 for review, except this one point.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (mitchmindtree) [reply nixos/modules/services/web-apps/nextcloud.nix:330]: Can do, it will also simplify things a little. Shall we also remove `adminpass` and `dbpass` in favour of their `*File` alternatives too?

Comment (aanderse) [reply nixos/modules/services/web-apps/nextcloud.nix:330]: Ideally yes. We should work toward that. A separate PR perhaps?

**Comment by aanderse** [comment]
Thank you @mitchmindtree!

**Comment by Ma27** [comment]
Yeah, I have it on my todo list, but it's kinda overloaded, so I'll have to request some patience here :)

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [inline nixos/modules/services/web-apps/nextcloud.nix:343]: Does that mean, it's for services that are not from Amazon, but speak the "S3 protocol"? Anyways, please give some more context here.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [inline nixos/modules/services/web-apps/nextcloud.nix:354]: We either use SSL or we don't, correct? :) 
(and are there any reasons for turning it off? It should be `true` at least).

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [inline nixos/modules/services/web-apps/nextcloud.nix:382]: I don't see why this is needed. You can just do

``` nix
{
  objectstore = {
    enable = mkEnableOption "...";
    /* all the other objectstore parameters */
  };
}
```

if you can only specify a single object store.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [inline nixos/modules/services/web-apps/nextcloud.nix:385]: Same here.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [inline nixos/modules/services/web-apps/nextcloud.nix:485]: Please do this in a separate commit. While the change itself seems correct, I'd prefer to always do it. That way maintainers don't get commits when bisecting for an issue with a lot of unrelated changes.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [inline nixos/modules/services/web-apps/nextcloud.nix:494]: With my suggestion from above we can drop this assertion entirely.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [inline nixos/modules/services/web-apps/nextcloud.nix:558]: Nix is lazily evaluated. As soon as you drop the unnecessary submodule above and work with a single `enable` option, you can use e.g. `s3.arguments` (which should default to `{}` then rather than `null`) here without this if-else. I guess the same applies to `class =`.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [inline nixos/modules/services/web-apps/nextcloud.nix:569]: You can use `lib.boolToString` here.

**Review by Ma27 — CHANGES_REQUESTED** [review]
Since you mentioned "non-Amazon" implementations, can we use such an implementation in a VM test? :)

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (mitchmindtree) [reply nixos/modules/services/web-apps/nextcloud.nix:343]: I'm unsure about this! My limited understanding is that in some implementations of "S3-compatible" storage a hostname is required, whereas apparently it is not on Amazon? I'd have no idea having avoided Amazon, but I know it is required for my digital ocean space at least.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (mitchmindtree) [reply nixos/modules/services/web-apps/nextcloud.nix:354]: The difference between `null` and us deciding on a default value is that in the case the value is `null`, we don't add a `use_ssl` line to the config at all and in turn fall back to whatever Nextcloud's default happens to be.

That said I understand if you'd prefer we try to encourage the use of SSL - I'm happy to change the default to `true`.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (mitchmindtree) [reply nixos/modules/services/web-apps/nextcloud.nix:382]: The reason why most config is behind `objectstore.s3.*` and not directly behind `objectstore.*` is because nextcloud supports multiple classes of object storage, each with its own set of unique configuration parameters. According to [their docs](https://docs.nextcloud.com/server/latest/admin_manual/configuration_files/primary_storage.html) nextcloud currently supports using S3, Openstack Swift V1 and V2 classes of object storage as primary storage. The current structure allows to avoid breaking downstream configurations when someone inevitably decides to add support for Openstack Swift or some other future object storage class that nextcloud might support.

W.r.t. using `nullOr`, I was unsure if you could have a submodule default to `{}` in the case that it has non-default options! I'm happy to remove the `nullOr` indirection and use `{}` as a default instead, it would certainly clean things up a little, though I must admit it's a bit strange to me that nix allows having a submodule type instance that is missing fields that have no defaults. I guess a submodule is just an attrset under the hood though.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (mitchmindtree) [reply nixos/modules/services/web-apps/nextcloud.nix:494]: This check will become useful in the case that options are added for the other nextcloud-supported classes of object storage as mentioned above, as according to the nextcloud docs a user should only be allowed to select a single object storage class to use as primary storage.

**Comment by mitchmindtree** [comment]
> Since you mentioned "non-Amazon" implementations, can we use such an implementation in a VM test? :)

Oooo I'm unsure if there is a self-hosted S3-compatible object storage type out there? I'm personally using digital ocean's implementation at the moment. I'd imagine that it *would* be possible to do this with the Openstack Swift class object storage once options are added for it, though I've no idea of what's involved in setting it up.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (mitchmindtree) [reply nixos/modules/services/web-apps/nextcloud.nix:343]: I should add that my comment here provides just about as much clarity as [the nextcloud docs](https://docs.nextcloud.com/server/latest/admin_manual/configuration_files/primary_storage.html#simple-storage-service-s3) do.

**Comment by Ma27** [comment]
Yeah fine. I guess I'll run some tests against an S3 bucket and if it works, we can merge as soon as we cleared up all the other points.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [reply nixos/modules/services/web-apps/nextcloud.nix:343]: Yeah, makes sense.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [reply nixos/modules/services/web-apps/nextcloud.nix:354]: > we don't add a use_ssl line to the config at all and in turn fall back to whatever Nextcloud's default happens to be

Can't we just use the fallback that Nextcloud uses (i.e. Nextcloud's default is our default). Having `null` always makes a special-case mandatory which isn't really nice.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [reply nixos/modules/services/web-apps/nextcloud.nix:382]: Then do `objectstore.s3.enable = mkEnableOption "...";`, please. A submodule is absolutely not needed here so far and just causes unnecessary complexity. If this really becomes necessary, we can switch to `types.attrsOf (types.submodule {})` sooner or later or implement the other obejctstore things manually if someone's willing to do that.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [reply nixos/modules/services/web-apps/nextcloud.nix:494]: This feels like premature optimization. As suggested above, `objectstore.s3.<option>` is good-enough for a tentative design, more things (including such an assertion) can be added later when we need it.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (mitchmindtree) [reply nixos/modules/services/web-apps/nextcloud.nix:354]: Sure, it's not documented anywhere but [judging by the src](https://github.com/nextcloud/server/blob/e2116e2fb226890341c548e3f7d79c0ac63c1b06/lib/private/Files/ObjectStore/S3ConnectionTrait.php#L83) it looks like SSL is enabled by default.

**Comment by mitchmindtree** [comment]
Okydoke @Ma27, I've rebased onto the latest master and added two more commits addressing your comments, along with another fixing the message of the exception that is thrown in the case that the specified secret file doesn't exist.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [inline nixos/modules/services/web-apps/nextcloud.nix:567]: PHP doesn't really care about indentation and this is not only weird to read, but also highly ambiguous because `'''` inside a multi-line string is actually `''` in the evaluated result.

**Review by Ma27 — COMMENTED** [review]
Apart from a minor nitpick, the change is looking good now.

I'd like to test it against an S3 bucket (I'll get to it today or later this week), then we can merge. When I check it out locally, I can fix the styling-issue I pointed out by myself (assuming you allow maintainers to push to this branch).

**Comment by mitchmindtree** [comment]
> assuming you allow maintainers to push to this branch

Is that what this refers to?

Screenshot from 2021-10-05 17-12-55

Either way, I've had a go at addressing the styling nit just in case.

**Review by Ma27 — APPROVED** [review]

**Pull request merged**
